src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1391 @@
+
+    /// Retrieve the allocation for operand `idx.1` at instruction `idx.0`
+    fn index(&self, idx: (usize, usize)) -> &Allocation {
+        &self.allocs[self.inst_alloc_offsets[idx.0] as usize + idx.1]
+    }
+}
+
+impl IndexMut<(usize, usize)> for Allocs {
+    fn index_mut(&mut self, idx: (usize, usize)) -> &mut Allocation {
+        &mut self.allocs[self.inst_alloc_offsets[idx.0] as usize + idx.1]
+    }
+}
+
+#[derive(Debug)]
+struct Stack<'a, F: Function> {
+    num_spillslots: u32,
+    func: &'a F,
+}
+
+impl<'a, F: Function> Stack<'a, F> {
+    fn new(func: &'a F) -> Self {
+        Self {
+            num_spillslots: 0,
+            func,
+        }
+    }
+
+    /// Allocates a spill slot on the stack for `vreg`
+    fn allocstack(&mut self, class: RegClass) -> SpillSlot {
+        trace!("Allocating a spillslot for class {class:?}");
+        let size: u32 = self.func.spillslot_size(class).try_into().unwrap();
+        // Rest of this function was copied verbatim
+        // from `Env::allocate_spillslot` in src/ion/spill.rs.
+        let mut offset = self.num_spillslots;
+        // Align up to `size`.
+        debug_assert!(size.is_power_of_two());
+        offset = (offset + size - 1) & !(size - 1);
+        let slot = if self.func.multi_spillslot_named_by_last_slot() {
+            offset + size - 1
+        } else {
+            offset
+        };
+        offset += size;
+        self.num_spillslots = offset;
+        trace!("Allocated slot: {slot}");
+        SpillSlot::new(slot as usize)
+    }
+}
+
+#[derive(Debug)]
+struct Edits {
+    /// The final output edits.
+    edits: Vec<(ProgPoint, Edit)>,
+    fixed_stack_slots: PRegSet,
+    /// The scratch registers being used in the instruction being
+    /// currently processed.
+    scratch_regs: PartedByRegClass<Option<PReg>>,
+    dedicated_scratch_regs: PartedByRegClass<Option<PReg>>,
+}
+
+impl Edits {
+    fn new(
+        fixed_stack_slots: PRegSet,
+        max_operand_len: u32,
+        num_insts: usize,
+        dedicated_scratch_regs: PartedByRegClass<Option<PReg>>,
+    ) -> Self {
+        // Some operands generate edits and some don't.
+        // The operands that generate edits add no more than two.
+        // Some edits are added due to clobbers, not operands.
+        // Anyways, I think this may be a reasonable guess.
+        let inst_edits_len_guess = max_operand_len as usize * 2;
+        let total_edits_len_guess = inst_edits_len_guess * num_insts;
+        Self {
+            edits: Vec::with_capacity(total_edits_len_guess),
+            fixed_stack_slots,
+            scratch_regs: dedicated_scratch_regs.clone(),
+            dedicated_scratch_regs,
+        }
+    }
+}
+
+impl Edits {
+    fn is_stack(&self, alloc: Allocation) -> bool {
+        if alloc.is_stack() {
+            return true;
+        }
+        if alloc.is_reg() {
+            return self.fixed_stack_slots.contains(alloc.as_reg().unwrap());
+        }
+        false
+    }
```

Comment:
We can write this as `alloc.is_stack() || (alloc.is_reg() && self.fixed_stack_slots.contains(...))` -- possibly a little more concise?